### PR TITLE
[Merged by Bors] - chore(RingTheory/PrincipalIdealDomain): PID implies Noetherian

### PR DESCRIPTION
**Pull request by pechersky**
Also, generalize `IsPrincipal.generator` to `Semiring R` and `AddCommMonoid M`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
I thought we already have https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/PrincipalIdealDomain.html#PrincipalIdealRing.isNoetherianRing

**Comment by pechersky** [comment]
> I thought we already have https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/PrincipalIdealDomain.html#PrincipalIdealRing.isNoetherianRing

Ah, yes. I'll generalize/rename.

File: Mathlib/RingTheory/PrincipalIdealDomain.lean [hunk 1]
```diff
@@ -103,7 +103,18 @@ theorem mem_iff_eq_smul_generator (S : Submodule R M) [S.IsPrincipal] {x : M} :
 theorem eq_bot_iff_generator_eq_zero (S : Submodule R M) [S.IsPrincipal] :
     S = ⊥ ↔ generator S = 0 := by rw [← @span_singleton_eq_bot R M, span_singleton_generator]
 
-end Ring
+protected lemma FG {S : Submodule R M} (h : S.IsPrincipal) : S.FG :=
+  ⟨{h.generator}, by simp only [Finset.coe_singleton, span_singleton_generator]⟩
+
+instance _root_.PrincipalIdealRing.isNoetherianRing [IsPrincipalIdealRing R] :
```
Comment (erdOne) [inline Mathlib/RingTheory/PrincipalIdealDomain.lean:109]: Please keep the lowered priority.

File: Mathlib/RingTheory/PrincipalIdealDomain.lean [hunk 2]
```diff
@@ -103,7 +103,18 @@ theorem mem_iff_eq_smul_generator (S : Submodule R M) [S.IsPrincipal] {x : M} :
...
+    IsNoetherianRing R where
+  noetherian S := (IsPrincipalIdealRing.principal S).FG
+
+instance _root_.IsPrincipalIdealRing.of_isNoetherianRing_of_isBezout
```
Comment (erdOne) [inline Mathlib/RingTheory/PrincipalIdealDomain.lean:113]: Also here

File: Mathlib/RingTheory/PrincipalIdealDomain.lean (hunk 1 above)
Comment (pechersky) [reply Mathlib/RingTheory/PrincipalIdealDomain.lean:109]: Sure, but what effect does this have for prop valued instances? lean4 handles TC loops.

Comment (erdOne) [reply Mathlib/RingTheory/PrincipalIdealDomain.lean]: The thing is if you have some `instance : Foo X`, then everytime the instance search sees a goal of type `Foo (some term)`,  this lemma will trigger, where usually we want the `Foo` to be derived from the structure of `some term`. See note [lower instance priority] for details.

**Comment by erdOne** [comment]
maintainer merge

File: Mathlib/RingTheory/PrincipalIdealDomain.lean [hunk 3]
```diff
@@ -103,7 +103,20 @@ theorem mem_iff_eq_smul_generator (S : Submodule R M) [S.IsPrincipal] {x : M} :
 theorem eq_bot_iff_generator_eq_zero (S : Submodule R M) [S.IsPrincipal] :
     S = ⊥ ↔ generator S = 0 := by rw [← @span_singleton_eq_bot R M, span_singleton_generator]
 
-end Ring
+protected lemma FG {S : Submodule R M} (h : S.IsPrincipal) : S.FG :=
```
Comment (jcommelin) [inline Mathlib/RingTheory/PrincipalIdealDomain.lean:106]: Elsewhere in the library we use `fg` in lemma names...
```suggestion
protected lemma fg {S : Submodule R M} (h : S.IsPrincipal) : S.FG :=
```

File: Mathlib/RingTheory/PrincipalIdealDomain.lean [hunk 4]
```diff
@@ -103,7 +103,20 @@ theorem mem_iff_eq_smul_generator (S : Submodule R M) [S.IsPrincipal] {x : M} :
...
+  ⟨{h.generator}, by simp only [Finset.coe_singleton, span_singleton_generator]⟩
+
+-- See note [lower instance priority]
+instance (priority := 100) _root_.PrincipalIdealRing.isNoetherianRing [IsPrincipalIdealRing R] :
+    IsNoetherianRing R where
+  noetherian S := (IsPrincipalIdealRing.principal S).FG
```
Comment (jcommelin) [inline Mathlib/RingTheory/PrincipalIdealDomain.lean:112]:
```suggestion
  noetherian S := (IsPrincipalIdealRing.principal S).fg
```

**Review by jcommelin — COMMENTED** [review]
Thanks!

bors d+

**Pull request closed without merging**